src/Elastic.Apm/Api/ITransaction.cs

```diff
@@ -50,5 +50,11 @@ public interface ITransaction : IExecutionSegment
 		/// existing one.
 		/// </returns>
 		string EnsureParentId();
+
+		/// <summary>
+		/// If true, then the transaction name was modified by external code, and transaction name should not be changed
+		/// or "fixed" automatically.
+		/// </summary>
+		bool HasCustomName { get; }
```

Comment:
Why do we move this to the public `ITransaction`? 

Asking because I'd think that public API users would not really care about this flag, but from this point it'll show up for them as well.

Reply:
I moved this to the interface because the ElasticApmModule needs to know whether it can rename a transaction name in an ASP.NET Full Framework app, similar to ASP.NET Core, and the started transaction returns the `ITransaction` interface: https://github.com/elastic/apm-agent-dotnet/blob/5f9fd02061e5eb2b83f839f51e5d47d2786587f3/src/Elastic.Apm.AspNetFullFramework/ElasticApmModule.cs#L137

I did consider

1. casting the `ITransaction` to `Transaction`, which seems like it might be the wrong thing to do as users may implement their own `ITransaction` type?
2. instantiating a `Transaction` directly, but this also seems like it might be wrong because it circumvents the public interface?

Another approach here may be to encapsulate whether renaming is allowed inside of the `Transaction.Name` setter e.g.

https://github.com/elastic/apm-agent-dotnet/blob/a63b7575469659d4a83a67da5ec0a613e8ec0a0a/src/Elastic.Apm.AspNetCore/WebRequestTransactionCreator.cs#L183-L202

would be something like

```c#
internal class Transaction : ITransaction
{
        [JsonIgnore]
	internal bool HasCustomName { get; private set; }

	[JsonConverter(typeof(TrimmedStringJsonConverter))]
	public string Name
	{
		get => _name;
		set
		{
			if (!HasCustomName)
			{
				HasCustomName = true;
				_name = value;
			}
		}
	}
}
```

```c#
//fixup Transaction.Name - e.g. /user/profile/1 -> /user/profile/{id}
var routeData = context.GetRouteData()?.Values;

if (routeData != null && context.Response.StatusCode != StatusCodes.Status404NotFound)
{
	logger?.Trace()?.Log("Calculating transaction name based on route data");
	var name = GetNameFromRouteContext(routeData);

	if (!string.IsNullOrWhiteSpace(name)) transaction.Name = $"{context.Request.Method} {name}";
}
else if (context.Response.StatusCode == StatusCodes.Status404NotFound)
{
	logger?.Trace()
		?
		.Log("No route data found or status code is 404 - setting transaction name to 'unknown route");
	transaction.Name = $"{context.Request.Method} unknown route";
}
``` 

The downside is that the work to calculate a new transaction name is performed whether or not it can be updated, and if a new name is not assigned, there is no indication. The first issue could be achieved by introducing an `ITransaction.SetName(Func<string> func)`, where the delegate is run only if a new name can be applied, and the second issue could be signalled by logging.

Reply:
Discussed over zoom - but for keeping it also here: I think best would be to check if it's `Transaction` if so, then cast, otherwise ignore. 

The idea behind having `ITransaction` is that potentially we have `NoopTransaction` when the agent is deactivated - and in that case if we ignore this field is fine, since nothing will be sent anyway.